File: LeetCode/Medium/1498-Number_Of_Subsequences_That_Satisfy_The_Given_Sum_Condition/1498-NumberOfSubsequencesThatSatisfyTheGivenSumCondition.js
const MOD = 1_000_000_007;
const MAX_N = 100_000;

let powers = new Array(MAX_N);

function precomputePowers() {
    powers[0] = 1;

    for (let i = 1; i <= MAX_N; i++) {
        powers[i] = (powers[i - 1] * 2) % MOD;
    }
}

precomputePowers();

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var numSubseq = function (nums, target) {
    let count = 0;

    nums.sort((a, b) => a - b);

    let left = 0;
    let right = nums.length - 1;

    while (left <= right) {
        if (nums[left] + nums[right] <= target) {
            count += powers[right - left];
            left++;
        } else {
            right--;
        }
    }

    return count % MOD;
};

File: LeetCode/Medium/1498-Number_Of_Subsequences_That_Satisfy_The_Given_Sum_Condition/1498-NumberOfSubsequencesThatSatisfyTheGivenSumCondition.cpp
const int MOD=1e9+7;
const int MAX_N=1e5;

std::vector<long long> powers(MAX_N);

void precompute_powers() {
    powers[0] = 1;

    for (int i = 1; i < MAX_N; i++) {
        powers[i] = (powers[i-1] * 2) % MOD;
    }
}

struct PrecomputationInitializer {
    PrecomputationInitializer() {
        precompute_powers();
    }
};

static PrecomputationInitializer initializer;

class Solution {
public:
    int numSubseq(vector<int>& nums, int target) {
        long long count = 0;

        std::sort(nums.begin(), nums.end());

        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            if (nums[left] + nums[right] <= target) {
                count += powers[right - left];
                left++;
            } else {
                right--;
            }
        }

        return count % MOD;
    }
};